kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,118 @@ struct LogReplayScanner {
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            true
+        } else {
+            debug!(
+                "Including ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            if self.is_log_batch {
+                // Remember file actions from this batch so we can ignore duplicates as we process
+                // batches from older commit and/or checkpoint files. We don't track checkpoint
+                // batches because they are already the oldest actions and never replace anything.
+                self.seen.insert(key);
+            }
+            false
         }
     }
+
+    /// True if this row contains an Add action that should survive log replay. Skip it if the row
+    /// is not an Add action, or the file has already been seen previously.
+    fn is_valid_add<'a>(&mut self, i: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<bool> {
+        // Add will have a path at index 0 if it is valid; otherwise, if it is a log batch, we
+        // may have a remove with a path at index 4.
+        let (path, getters, is_add) = if let Some(path) = getters[0].get_str(i, "add.path")? {
+            (path, &getters[1..4], true)
+        } else if !self.is_log_batch {
+            return Ok(false);
+        } else if let Some(path) = getters[4].get_opt(i, "remove.path")? {
+            (path, &getters[5..8], false)
+        } else {
+            return Ok(false);
+        };
+
+        let dv_unique_id = match getters[0].get_opt(i, "deletionVector.storageType")? {
+            Some(storage_type) => Some(DeletionVectorDescriptor::unique_id_from_parts(
+                storage_type,
+                getters[1].get(i, "deletionVector.pathOrInlineDv")?,
+                getters[2].get_opt(i, "deletionVector.offset")?,
+            )),
+            None => None,
+        };
+
+        // Process both adds and removes, but only return not already-seen adds
+        let file_key = FileActionKey::new(path, dv_unique_id);
+        Ok(!self.already_seen(file_key) && is_add)
+    }
 }
 
-impl RowVisitor for AddRemoveVisitor {
+impl<'seen> RowVisitor for AddRemoveDedupVisitor<'seen> {
     fn selected_column_names_and_types(&self) -> (&'static [ColumnName], &'static [DataType]) {
-        // NOTE: The visitor assumes adds always come first, with removes optionally afterward
-        static ALL_NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> = LazyLock::new(|| {
-            let (add_names, add_fields) = AddVisitor::names_and_types();
-            let (remove_names, remove_fields) = RemoveVisitor::names_and_types();
-            let names = add_names.iter().chain(remove_names).cloned().collect();
-            let fields = add_fields.iter().chain(remove_fields).cloned().collect();
-            (names, fields).into()
+        // NOTE: THe visitor assumes adds always come first, with removes optionally afterward
```

Comment:
```suggestion
        // NOTE: The visitor assumes adds always come first, with removes optionally afterward
```

Reply:
also might be worth clarifying: "comes first" as in it comes first in the schema, not in the log, right? iirc, you can't assume any ordering among actions within a commit.